src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala

```diff
@@ -76,4 +78,65 @@ private[actions] abstract class RefreshActionBase(
           s"Current index state is ${previousIndexLogEntry.state}")
     }
   }
+
+  /**
+   * Compare list of source data files from previous IndexLogEntry to list
+   * of current source data files, validate fileInfo for existing files and
+   * identify deleted source data files. Finally, append the previously known
+   * deleted files to the result.
+   */
+  protected lazy val deletedFiles: Seq[String] = {
+    val rels = previousIndexLogEntry.relations
+    val originalFiles: Set[FileInfo] = rels.head.data.properties.content.fileInfos
+    val currentFiles: Map[String, FileInfo] = rels.head.rootPaths
+      .flatMap { p =>
+        Content
+          .fromDirectory(path = new Path(p), throwIfNotExists = true)
+          .fileInfos
+      }
+      .map(f => f.name -> f)
+      .toMap
+
+    var delFiles = Seq[String]()
+    originalFiles.foreach { f =>
+      currentFiles.get(f.name) match {
+        case Some(v) =>
+          if (!f.equals(v)) {
+            throw HyperspaceException(
+              "Index refresh (to handle deleted source data) aborted. " +
+                s"Existing source data file info is changed (file: ${f.name}).")
+          }
+        case None => delFiles :+= f.name
+      }
+    }
+
+    delFiles ++ previousIndexLogEntry.deletedFiles
+  }
+
+  /**
+   * Compare list of source data files from previous IndexLogEntry to list
+   * of current source data files, validate fileInfo for existing files and
+   * identify newly appended source data files. Finally, append the previously known
+   * appended files to the result.
+   */
+  protected lazy val appendedFiles: Seq[String] = {
+    val relation = previousIndexLogEntry.relations.head
+    val originalFiles = relation.data.properties.content.fileInfos
+
+    val currentFiles = df.queryExecution.optimizedPlan
```

Comment:
Could you extract this as `protected lazy val` and use it for both appendedFiles and deletedFiles?